docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,269 @@
+* We can create a similar library wrapper imitating [Bull](https://github.com/OptimalBits/bull) using this broker as backend
+* Using message priority we can have large back ground scripts generate a lot of event which will not hamper user action generating events from application
+
+## Key Concepts
+
+| Concept | Definition |
+| -- | -- |
+| Channel | A channel is the broadcast highway for messages. |
+| Producer | Producer is the system generating a message and broadcasting it to a channel |
+| Consumer | Consumer is the system registering itself to listen to messages broadcasted to a specific channel |
+| DLQ / Dead Letter Queue | A queue (collection) of messages that failed deliver to the consumer for non 2XX response from client |
+| Message | The payload Producer wants distributed across the Consumer within a Channel |
+| Message Delivery | Reaching the consumer with the message should be considered as delivery, in case of success it would be marked as consumed, else end up in dead letter queue |
+| Message Status | Message has 2 status - Acknowledged and Out-for-delivery |
+| Message Delivery Status | Message Delivery status is associated with a Message, Channel and Consumer combination and is a enumeration of - In-flight, Retry-Delivery, Retry-In-flight, Delivered, Dead |
+| Rational-delay | Time delta for fail-safe mechanism to kick in |
+| Delivery-Timeout | How long will we wait for message delivery to finish before we step in |
+
+## Life-cycle of a Message
+
+### High-level Flow
+
+* Broker receives a **Message** addressed to a **Channel** from a **Producer**
+* Broker retrieves the **Consumers** interested in broadcast of a **Channel**
+* Broker _delivers_ **Message** to each **Consumer** retrieved in previous step
+
+### Key milestones in that flow
+
+* Acknowledge to **Producer** of receipt of **Message**
+* Track _delivery status_ of the **Message** against each **Consumer**
+
+### Non-trivial scenarios
+
+* Broker crashed before sending the acknowledgement
+* Broker crashed after sending the acknowledgement but before attempting to deliver the messages
+* Broker crashed attempting to deliver to some/all consumers but succeeded with others (if any)
+
+### Design considerations
+
+Based on the flow, non-trivial scenarios we want to derive at the following design considerations -
+
+* Acknowledgement should be tied to being able to store Message **and** sending the ACK signal, but not dependent on Producer received the ACK signal.
+* There should be a fail-safe way to ensure Acknowledged messages are attempted to deliver in case synchronous triggering of delivery process fails
+* Consumer may not be available, so we should have a retry policy
+
+### One level deeper
+
+So based on the above conversation lets elaborate the high-level flow to more concrete blocks and dive deep in them. The building blocks would be -
+
+1. Receive Message
+1. Start Delivery Process
+1. Attempt Individual Delivery
+
+Lets look into each of them.
```

Comment:
I think it would be helpful if you had some kind of flow diagram to clarify the message receipt and delivery protocols because I'm having a hard time piecing it together from the definitions below.

Reply:
Good point, I will try to come up with a sequence diagram.